src/hotspot/share/classfile/javaClasses.hpp

```diff
@@ -137,11 +149,26 @@ class java_lang_String : AllStatic {
   static inline void set_value_raw(oop string, typeArrayOop buffer);
   static inline void set_value(oop string, typeArrayOop buffer);
 
+  // Set the no_deduplication flag true.  This flag is sticky; once set it
+  // never gets cleared.  This is set when a string is interned in the
+  // StringTable, to prevent string deduplication from changing the string's
+  // value array.
+  static inline void set_no_deduplication(oop java_string);
+
+  // Test and set the deduplication_requested flag.  Returns true if the
+  // flag's value is changed, false if it was already set.  This flag is
+  // sticky; once set it never gets cleared.  A GC may use this operation to
+  // determine whether to request deduplication of a string, avoiding
+  // multiple requests for the same string.
+  static inline bool set_deduplication_requested(oop java_string);
+
   // Accessors
   static inline typeArrayOop value(oop java_string);
   static inline typeArrayOop value_no_keepalive(oop java_string);
   static inline bool hash_is_set(oop string);
   static inline bool is_latin1(oop java_string);
+  static inline bool no_deduplication(oop java_string);
```

Comment:
"may_deduplicate" would require internally flipping the sense, to account for the initial value being false because of zero-initialization.  I've changed it to "deduplication_forbidden"; hopefully that helps.